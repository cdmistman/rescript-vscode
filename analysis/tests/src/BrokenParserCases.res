// I would want this to recover to {someProp: %rescript.exprhole, ...}, but it currently eats the tuple + parses to {someProp: otherProp}
// let _ = ({someProp: , otherProp: 123}, 123)
//                    ^ast

// Same problem here, does not recover from the missing prop, and ends up eating the rest of the record declaration while only parsing {firstProp: secondProp}
// let _ = someFunc({firstProp: , secondProp: 123, thirdProp: 123})
//                             ^ast

// I'd like the commas to be parsed as (broken) tuple "members" too. There's currently no robust way to derive _where_ in that tuple the cursor is, or even how many members the tuple has, because there's no pos match, and the empty comma item is just discarded.
// switch v { | (_, , _) => () }
//                 ^ast

// Same problem here, does not recover from the missing prop, and ends up eating the rest of the record declaration while only parsing {firstProp: secondProp}
// let {someField, , otherField, } = someVal
//                ^ast

