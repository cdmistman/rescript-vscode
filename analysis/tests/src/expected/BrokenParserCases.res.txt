Dump AST src/BrokenParserCases.res 1:22

Source:
// let _ = ({someProp: , otherProp: 123}, 123)
//                    ^ast

<*>Pstr_value(
  value:
    Ppat_any
  expr:
    <*>Pexp_record(
      fields:
        someProp: Pexp_ident:otherProp
    )
)

Dump AST src/BrokenParserCases.res 5:31

Source:
// let _ = someFunc({firstProp: , secondProp: 123, thirdProp: 123})
//                             ^ast

<*>Pstr_value(
  value:
    Ppat_any
  expr:
    <*>Pexp_apply(
      expr:
        Pexp_ident:someFunc
      args:
        <unlabelled>=
          <*>Pexp_record(
            fields:
              firstProp: Pexp_ident:secondProp
          ),
        <unlabelled>=
          Pexp_constant(Pconst_integer(123)),
        <unlabelled>=
          <x><unimplemented_pexp_desc: Pexp_constraint>,
        <unlabelled>=
          Pexp_constant(Pconst_integer(123))
    )
)

Dump AST src/BrokenParserCases.res 9:19

Source:
// switch v { | (_, , _) => () }
//                 ^ast

<*>Pstr_eval(
<*>Pexp_match(Pexp_ident:v)
    case 1:
      pattern<*>:
        <*>Ppat_tuple(
        Ppat_any,
        Ppat_any
    )
      expr:
        Pexp_construct(())
)

